doc/source/dev/database_session_management.md

```diff
@@ -0,0 +1,34 @@
+# How SQLAlchemy sessions are handled in Galaxy
+
+## Access to `sessionmaker`
+
+We have a central resource from which to grab SQLAlchemy's [sessionmaker](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.sessionmaker), that is `app.model.session`.
+You can either use it directly as a proxy to the current session or call `app.model.session()` to get the proxied Session object, or use the context manager (e.g. `with app.model.session() as session:`), as detailed in the SQLAlchemy documentation.
+
+## `sessionmaker` scopes
+
+The default scope of the sessionmaker is [thread local](https://docs.sqlalchemy.org/en/20/orm/contextual.html#thread-local-scope). This means that accessing `sessionmaker` will return the same session in the same thread. This default scope can be changed if necessary.
+
+### WSGI, ASGI, Celery
+
+In the context of the webapp (WSGI and ASGI) there is a clearly delineated lifespan for a session, and that lifespan isn't tied to a thread. so we set the scope using `app.model.set_request_id(request_id)` and we close and remove the session at the end of the request with `app.model.unset_request_id(request_id)`. Similarly, we set the scope for each Celery task execution using `set_request_id(request_id)` and `unset_request_id`.
+
+By using this mechanism we are certain that web requests and celery tasks receive a new session object and that any resources held by the session are closed and rolled back.
+
+### Job and Workflow handlers
+
+Job and Workflow handlers perform multiple pieces of business logic in separate threads and care needs to be taken when deciding when to start a transaction, when to do a rollback and if and where a session lifespan can be introduced.
+
+The following paragraph discusses the choices made in <https://github.com/galaxyproject/galaxy/blob/064360dab053426ce166bb7274d6a3a5790eb7cc/lib/galaxy/jobs/runners/__init__.py#L835-L843>:
+
+The `JobHandlerQueue.monitor` method is executed as a thread within the job handler process,
+and all work that requires access to the database is happening inside `check_watched_items`,
+which iterates over jobs assigned to this handler process. By setting the scope to a new,
+random uuid around `check_watched_items` we ensure that even if the session becomes corrupt
+the session is properly discarded and the following iteration of `check_watched_items` receives
```

Comment:
```suggestion
the session is properly discarded and the next iteration of `check_watched_items` receives
```
To me,  "the following iteration" reads as if it's pointing to more details about it in the next sentence.